DirectProgramming/C++SYCL_FPGA/ReferenceDesigns/matmul/src/matmul.hpp

```diff
@@ -0,0 +1,139 @@
+#ifndef __MATMUL_HPP__
+#define __MATMUL_HPP__
+
+#include <chrono>
+#include <iostream>
+
+#include <sycl/ext/intel/ac_types/ac_int.hpp>
+#include <sycl/ext/intel/fpga_extensions.hpp>
+#include <sycl/ext/intel/prototype/interfaces.hpp>
+#include <sycl/sycl.hpp>
+
+#include "memory_transfers.hpp"
+#include "streaming_matmul.hpp"
+
+using sycl::ext::intel::experimental::property::usm::buffer_location;
+
+// Forward declare the kernel and pipe names
+// (This prevents unwanted name mangling in the optimization report.)
+class FeederA;
+class FeederB;
+class Matmul;
+class Drain;
+class APipe;
+class BPipe;
+class CPipe;
+class DPipe;
+
+/**
+ * Implementation of the matrix multiplication using multiple streaming kernels.
+ * Parameterized by datatype, matrix size, and tile size. Exercises the kernels
+ * by running multiple repetitions for a set of matrices.
+ *
+ * Function arguments:
+ *  q: device queue
+ *  a_matrix: input matrix pointer (given in column-major)
+ *  b_matrix: input matrix pointer (given in row-major, i.e., transposed)
+ *  c_matrix: output matrix pointer (will be stored in column-major)
+ *  repetitions: number of repetitions of the computation to execute
+ *
+ */
+template <typename TT,          // Datatype of the elements of the matrix
+          int k_rows_a,         // Rows of matrix A
+          int k_common,         // Columns of matrix A / rows of matrix B
+          int k_cols_b,         // Columns of matrix B
+          int k_tile_a,         // Tile size for matrix A
+          int k_tile_b,         // Tile size for matrix B
+          int k_num_matrices>   // Number of pairs of matrices to multiply
+void MatmulImpl(sycl::queue &q,            // Device queue
+                std::vector<TT> &a_matrix, // Input matrix A
+                std::vector<TT> &b_matrix, // Input matrix B
+                std::vector<TT> &c_matrix, // Output matrix C = A * B
+                int repetitions            // Number of repetitions
+) {
+  // Number of elements per DDR access
+  // NOTE: optimized for single-precision floating-point matrices
+  constexpr int kElemsPerDDRAccess = 8;
+
+  // Matrix sizes
+  constexpr int kMatsizeA = k_rows_a * k_common;
+  constexpr int kMatsizeB = k_cols_b * k_common;
+  constexpr int kMatsizeC = k_rows_a * k_cols_b;
+
+  // Buffer locations for mmhost interfaces
```

Comment:
right now, these buffer locations are "needed" even when compiling with a BSP since the functors have "aspace" as a template parameter (even though that parameter is not used in the full-system flow). alternatively I could change the signature of the functor based on IPA/full-system, but would that be too messy?

Reply:
Don't worry about it then, I did not realize that they were used in the mem transfer